src/meta/src/hummock/manager/mod.rs

```diff
@@ -901,14 +892,24 @@ where
         };
         let compaction = compaction_guard.deref_mut();
         let start_time = Instant::now();
-        let mut compact_status = VarTransaction::new(
-            compaction
-                .compaction_statuses
-                .get_mut(&compact_task.compaction_group_id)
-                .ok_or(Error::InvalidCompactionGroup(
-                    compact_task.compaction_group_id,
-                ))?,
+        let compaction_groups: HashSet<_> = HashSet::from_iter(
+            self.compaction_group_manager
+                .compaction_group_ids()
+                .await
+                .into_iter(),
         );
+        let original_keys = compaction.compaction_statuses.keys().cloned().collect_vec();
+        let mut compact_statuses = BTreeMapTransaction::new(&mut compaction.compaction_statuses);
+        for group_id in original_keys {
+            if !compaction_groups.contains(&group_id) {
+                compact_statuses.remove(group_id);
```

Comment:
Directly remove compact_status may fail compaction heartbeat cancellation
https://github.com/risingwavelabs/risingwave/blob/cf5852f248735095f54f2b971153fbf4f0ba2a32/src/meta/src/hummock/manager/mod.rs#L249

which cleans up compaction heartbeat in
https://github.com/risingwavelabs/risingwave/blob/cf5852f248735095f54f2b971153fbf4f0ba2a32/src/meta/src/hummock/manager/mod.rs#L989

So when removing compact_status, we need also consider compact_task_assignment and compaction heartbeat.

Reply:
Similar here
https://github.com/risingwavelabs/risingwave/blob/cf5852f248735095f54f2b971153fbf4f0ba2a32/src/meta/src/hummock/manager/context.rs#L52

This call will be retried until success. But it can fail forever due to group's compact_status absence.

Reply:
> Directly remove compact_status may fail compaction heartbeat cancellation
> 
> https://github.com/risingwavelabs/risingwave/blob/cf5852f248735095f54f2b971153fbf4f0ba2a32/src/meta/src/hummock/manager/mod.rs#L249
> 
> which cleans up compaction heartbeat in
> 
> https://github.com/risingwavelabs/risingwave/blob/cf5852f248735095f54f2b971153fbf4f0ba2a32/src/meta/src/hummock/manager/mod.rs#L989
> 
> So when removing compact_status, we need also consider compact_task_assignment and compaction heartbeat.

fixed

Reply:
> Similar here
> 
> https://github.com/risingwavelabs/risingwave/blob/cf5852f248735095f54f2b971153fbf4f0ba2a32/src/meta/src/hummock/manager/context.rs#L52
> 
> This call will be retried until success. But it can fail forever due to group's compact_status absence.

fixed